[CIR] Upstream TernaryOp

mmha opened the pull request:
This patch adds TernaryOp to CIR plus a pass that flattens the operator in FlattenCFG.

This is the first PR out of (probably) 3 wrt. TernaryOp. I split the patches up to make reviewing them easier. As such, this PR is only about adding the CIR operation. The next PR will be about the CodeGen bits from the C++ conditional operator and the final one will add the cir-simplify transform for TernaryOp and SelectOp.

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -609,8 +609,8 @@ def ConditionOp : CIR_Op<"condition", [
 def YieldOp : CIR_Op<"yield", [ReturnLike, Terminator,
-                               ParentOneOf<["IfOp", "ScopeOp", "WhileOp",
-                                            "ForOp", "DoWhileOp"]>]> {
+                               ParentOneOf<["IfOp", "TernaryOp", "ScopeOp",
```
What is the ordering of this list?  Does it make sense to group them in some way?  Or perhaps alpha order?

mmha replied:
I put it next to `IfOp` since both are about branching. But maybe lexicographic ordering is more sensible. The incubator does neither FWIW.

erichkeane replied:
Probably not actionable on this patch, but something for @bcardosolopes to consider?  Perhaps an incubator patch to give SOME meaningful sort here :)

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -1246,6 +1246,59 @@ def SelectOp : CIR_Op<"select", [Pure,
+    operation. The first argument is a `cir.bool` condition to evaluate, followed
+    by two regions to execute (true or false). This is different from `cir.if`
+    since each region is one block sized and the `cir.yield` closing the block
+    scope should have one argument.
```
Do we represent the GNU extension of `x = Thing1 ?: Thing2;` any differently?

mmha replied:
GNU `?:` is only partially implemented (we lack glvalue support for opaque value expressions) but it's represented by `cir.ternary`.

erichkeane replied:
Can we document it here?  One of the 'features' of it IIRC is that it only evaluates `Thing1` 1x instead of 2x (like in a traditional ternary), so it might be nice to document how that is represented here.

erichkeane reviewed (commented):
Not the best one to do the full review here, but had 2 comments.

bcardosolopes reviewed (approved):
> I split the patches up to make reviewing them easier

Way to go, thanks!

LGTM, after doc changes requested by Erich

andykaylor commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
````diff
@@ -1246,6 +1246,59 @@ def SelectOp : CIR_Op<"select", [Pure,
+    Example:
+
+    ```mlir
+    // x = cond ? a : b;
````
```suggestion
    // cond = a && b;
```
Along with corresponding changes below. (See https://godbolt.org/z/TjxY77xEW)

andykaylor commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -1246,6 +1246,59 @@ def SelectOp : CIR_Op<"select", [Pure,
+  let arguments = (ins CIR_BoolType:$cond);
+  let regions = (region AnyRegion:$trueRegion,
+                        AnyRegion:$falseRegion);
+  let results = (outs Optional<CIR_AnyType>:$result);
```
What does it mean for this to be optional? Can we have a TernaryOp that doesn't return a value?

mmha replied:
Yes, you can write this:

```c++
b ? (void) a : (void) b;
```

in which case either `a` or `b` get evaluated but the result of the ternary is discarded.

andykaylor commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -1058,6 +1058,48 @@ LogicalResult cir::BinOp::verify() {
+// TernaryOp
+//===----------------------------------------------------------------------===//
+
+/// Given the region at `index`, or the parent operation if `index` is None,
```
Should this be "the region at `point`"?

andykaylor commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -1058,6 +1058,48 @@ LogicalResult cir::BinOp::verify() {
+    return;
+  }
+
+  // If the condition isn't constant, both regions may be executed.
```
This comment seems to imply a constant check that isn't being performed.

andykaylor commented on clang/lib/CIR/Dialect/Transforms/FlattenCFG.cpp:
```diff
@@ -254,10 +254,61 @@ class CIRLoopOpInterfaceFlattening
+    // when relevant.
+    if (op->getResultTypes().size())
+      locs.push_back(loc);
+    auto *continueBlock =
```
```suggestion
    Block *continueBlock =
```

andykaylor commented on clang/lib/CIR/Dialect/Transforms/FlattenCFG.cpp:
```diff
@@ -254,10 +254,61 @@ class CIRLoopOpInterfaceFlattening
+    falseBlock = &falseRegion.front();
+    mlir::Operation *falseTerminator = falseRegion.back().getTerminator();
+    rewriter.setInsertionPointToEnd(&falseRegion.back());
+    cir::YieldOp falseYieldOp = dyn_cast<cir::YieldOp>(falseTerminator);
```
```suggestion
    auto falseYieldOp = dyn_cast<cir::YieldOp>(falseTerminator);
```

mmha committed:
Apply suggestions from code review

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

mmha commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -609,8 +609,8 @@ def ConditionOp : CIR_Op<"condition", [
 def YieldOp : CIR_Op<"yield", [ReturnLike, Terminator,
-                               ParentOneOf<["IfOp", "ScopeOp", "WhileOp",
-                                            "ForOp", "DoWhileOp"]>]> {
+                               ParentOneOf<["DoWhileOp", "ForOp", "WhileOp",
```
See Erichs comment about my placement of `"TernaryOp"` in that list. I just created a PR in the incubator here: https://github.com/llvm/clangir/pull/1586

bcardosolopes replied:
Merged!

andykaylor replied:
WhileOP is still in the wrong place, though, right?

mmha commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -1246,6 +1246,59 @@ def SelectOp : CIR_Op<"select", [Pure,
+def TernaryOp : CIR_Op<"ternary",
+      [DeclareOpInterfaceMethods<RegionBranchOpInterface>,
+       RecursivelySpeculatable, AutomaticAllocationScope, NoRegionArguments]> {
+  let summary = "The `cond ? a : b` C/C++ ternary operation";
```
We simplify `cir.ternary` to `cir.select` in FlattenCFG. See the tests [here](https://github.com/llvm/clangir/blob/main/clang/test/CIR/CodeGen/ternary.cpp). Without `-O1` the first test case gets represented by `cir.ternary`: https://godbolt.org/z/fz39rh5Kn

andykaylor replied:
Oh. I didn't realize we ran CIR simplification based on the opt level. I guess it makes sense in a way, but I'm not thrilled with the idea that this is happening in the front end. This relates to the suggestion I made elsewhere that maybe we should be moving the lowering and transforms out of the clang component.

andykaylor reviewed (approved):
I've still got a couple of open nits about the comments in CIRDialect.cpp, but otherwise this looks good.

mmha committed:
[CIR] Upstream TernaryOp

This patch adds TernaryOp to CIR plus a pass that flattens the operator in FlattenCFG.

mmha committed:
Address feedback on comments

erichkeane reviewed (approved):
I'm happy once Andy is.
